### ISSUE TITLE
About the batch-level accumulated loss

### ISSUE THREAD

**Issue by ZhengyuZhao** [jeromerony/fast_adversarial#5]
Hi,

I am confused about the implementation in the 
line "ce_loss = F.cross_entropy(logits, labels, reduction='sum')" of ddn.py.

In principle, I think it should be separated losses computed for each image, i.e., "ce_loss = F.cross_entropy(logits, labels, reduction=='none')", instead of a single batch-level accumulated loss applied for each image in that batch, as you have described in your paper.

Of course, I suppose the results should be somehow very near.

**Comment by jeromerony** [comment]
Hi,

In PyTorch, you cannot call the autograd.grad (or the backward method) on a tensor. It can only be done implicitly on a scalar, hence the sum. From a math standpoint, this is perfectly equivalent.

If you do it image-by-image or with a sum, the only difference you might observe may come from numerical error.

**Comment by ZhengyuZhao** [comment]
> Hi,
> 
> In PyTorch, you cannot call the autograd.grad (or the backward method) on a tensor. It can only be done implicitly on a scalar, hence the sum. From a math standpoint, this is perfectly equivalent.
> 
> If you do it image-by-image or with a sum, the only difference you might observe may come from numerical error.

Thanks for your prompt reply.


1. For sure, you can call the autograd.grad (or the backward method) on a list of losses with respect to a list of the separate corresponding image inputs in that batch, which also benefits from the GPU acceleration.

2. I don't think they are perfectly equivalent, because using the mean (or sum) loss of a batch of images for every image is not the same as using their own losses for each image individually.
In the case of training a network, perhaps it doesn't matter, but for the adversarial attacks (in the testing stage), it should be something we should consider.

**Comment by jeromerony** [comment]
This would mean converting the tensors, to lists of tensors which is less practical for indexing.

They are equivalent with a **sum**, not with a mean. 
Using a mean would divide the gradient by the number of items in the list, which might be problematic for initial steps where the gradient is usually already small. With a sum, we have not observed any difference.

The best way to verify your claim is to test and report.

**Comment by ZhengyuZhao** [comment]
Thanks for your further explanations.
But I am sorry that I can't really get your point. And I am not concerning about the gradient vanishing caused by the mean operation.

Let's say we have 2 toy images x1 and x2,
And the losses for these two images are denoted as loss1 and loss2, respectively.
By using a sum, you only get a single loss_batch=loss1+loss2 ( or by a mean loss_batch=(loss1+loss2)/2),
then you use this loss_batch for x1 and also x2.
But in fact, it should be using loss 1 for x1, while loss2 for x2.
The updating of the loss_batch with respect to x1 will be definitely different from using loss1 for it ( the same case for x2). So the process for different images should be parallel.

I will try to reimplement it if you still think they are the same after my above description.

**Comment by luizgh** [comment]
@ZhengyuZhao please note that the gradient of `loss2` wrt to image `x1` is zero, since the losses for different inputs are independent (i.e. `x1` is not used in the graph to compute `loss2`, when the model is in "eval" model, otherwise batchnorm mixes the inputs). Therefore, the gradient of `loss = loss1 + loss2` wrt to x1 and `loss1` wrt to x1 are the same.

**Comment by ZhengyuZhao** [comment]
> @ZhengyuZhao please note that the gradient of `loss2` wrt to image `x1` is zero, since the losses for different inputs are independent (i.e. `x1` is not used in the graph to compute `loss2`, when the model is in "eval" model, otherwise batchnorm mixes the inputs). Therefore, the gradient of `loss = loss1 + loss2` wrt to x1 and `loss1` wrt to x1 are the same.

Many thanks, I think now I understand how the function works!
